File: fsharp/tournament/Tournament.fs
module Tournament
    
type Result =
    | Win
    | Lost
    | Draw

type Schedule ={
    Team: string;
    Wins: int;
    Losses: int;
    Draws: int;
    Points: int;
    }

let parse (line:string) =
    line.Split(";")
    |> function 
        | [|team1; team2; result |]-> 
            match result with
            | "win" -> [{Team = team1; Wins = 1; Losses = 0; Draws = 0; Points=3};{Team = team2; Wins = 0; Losses = 1; Draws = 0; Points =0;}]
            | "loss" -> [{Team = team1; Wins = 0; Losses = 1; Draws = 0; Points=0;};{Team = team2; Wins = 1; Losses = 0; Draws = 0; Points=3;}]
            | "draw" -> [{Team = team1; Wins = 0; Losses = 0; Draws = 1; Points=1;};{Team = team2; Wins = 0; Losses = 0; Draws = 1; Points=1;}]
            | _ -> failwithf "%A is unrecognized game result" result
        | _ -> failwithf "%A is not formatted as as <team>;<team>;<result>" line


let condenseSchedule team games = 
    let wins = games |> List.sumBy (fun g -> g.Wins)
    let losses = games |> List.sumBy (fun g -> g.Losses)
    let draws = games |> List.sumBy (fun g -> g.Draws)
    let points = games |> List.sumBy (fun g -> g.Points)
    {Team = team; Wins = wins; Losses = losses; Draws = draws; Points  = points;}

let prettyPrintTeam team =
    sprintf "%-30s |%3d |%3d |%3d |%3d |%3d" team.Team (team.Wins + team.Losses + team.Draws) team.Wins team.Draws team.Losses team.Points

let prettyPrintLeague schedule =
    "Team                           | MP |  W |  D |  L |  P" :: (schedule |> List.map prettyPrintTeam)

let tally input = 
    input 
    |> List.map parse 
    |> List.concat 
    |> List.groupBy (fun item -> item.Team) 
    |> List.map (fun (teams, games) -> condenseSchedule teams games)
    |> List.groupBy (fun schedule -> schedule.Points)
    |> List.sortByDescending (fun (points,_) -> points)
    |> List.map (fun (_, teams) -> teams |> List.sortBy( fun t -> t.Team))
    |> List.concat
    |> prettyPrintLeague

File: python/tournament/tournament.py
def add_win(record):
    wins, losses, draws, points = record
    return (wins+1, losses, draws, points +3)

def add_loss(record):
    wins, losses, draws, points = record
    return (wins, losses+1, draws, points)

def add_draw(record):
    wins, losses, draws, points = record
    return (wins, losses, draws+1, points+1)

def format(name, record):
    (wins, draws, losses, points) = record
    formatted =f'{name:<31}|{wins+draws+losses:3} |{wins:3} |{losses:3} |{draws:3} |{points:3}' 
    print(formatted)
    return formatted

def tally(rows):
    scores = dict()
    for row in rows:
        parts= row.split(';')
        t1 = parts[0]
        t2 = parts[1]
        r = parts[2]
        if t1 not in scores:
            t1_tally = (0,0,0,0)
            scores[t1] = t1_tally
        else:
            t1_tally = scores[t1]

        if t2 not in scores:
            t2_tally = (0,0,0,0)
            scores[t2] = t2_tally
        else:
            t2_tally = scores[t2]

        if r == 'win':
            scores[t1] = add_win(t1_tally)
            scores[t2] = add_loss(t2_tally)
        elif r == 'loss':
            scores[t1] = add_loss(t1_tally)
            scores[t2] = add_win(t2_tally)
        else:
            scores[t1] = add_draw(t1_tally)
            scores[t2] = add_draw(t2_tally)

    results = [(team, record) for team,record in scores.items()]
    results.sort(key= lambda k: (-k[1][3], k[0]), reverse=True)
    results = [format(team, record) for team,record in results]
    results.append("Team                           | MP |  W |  D |  L |  P")
    results.reverse()
    return results